### PR TITLE
autoinstall: Don't use snap env when invoking early and late commands

### PR DESCRIPTION
**Pull request by Chris-Peterson444**
LP [#2032961](https://pad.lv/2032961): We are not cleaning the environment provided to subprocesses correctly. When trying to invoke a program that should run outside of the snap context, the `LD_LIBRARY_PATH` variable is pointing to the linker on the snap, which causes errors like:
```
curl: /lib/x86_64-linux-gnu/libc.so/6: version `GLIBC_2.33` not found (required by /snap/subiquity/x1/usr/lib/x86_64-linux-gnu/libpsl.so.5)
```

We had a similar issue solved in #1300. This fix similarly uses `orig_environ` to clean the environment before running commands in the respective command controllers. 

I would especially appreciate some input if the license header on the new test file looks correct. It's an exact copy and paste from one of the other new tests files from this year, but I just want to be sure.

### CODE REVIEW

File: subiquity/server/controllers/cmdlist.py [hunk 1]
```diff
@@ -124,7 +127,7 @@ def __init__(self, app):
                 ]
 
     def env(self):
-        env = super().env()
+        env = orig_environ(None)
```
Comment (ogayot) [inline subiquity/server/controllers/cmdlist.py:130]: Our [docs](https://github.com/canonical/subiquity/blob/9bf0a50a24b08ba7f80277212acf10a380ef03b0/documentation/autoinstall-reference.md#late-commands) indicate:

> You can run curtin in-target -- $shell_command [...] to run in the target system

and I think sadly your change will break the ability to do that. Maybe we need to update the docs, or add support for specifying if we want a clean env or not, as in:

```yaml
late-commands:
    # specifies a command directly (runs with a clean env)
    - echo "hello world"
    # Run `cat /etc/machine-id` in the target system
    - cmd: curtin in-target -- cat /etc/machine-id
      clean-env: false  # defaults to true? (true feels more natural but false might be better for backward compatibility)
``` 

or maybe we add the ability to run a command in the target system:

```yaml
late-commands:
    # specifies a command directly (runs in the installer environment)
    - echo "hello world"
    # Run `cat /etc/machine-id` in the target system
    - cmd: cat /etc/machine-id
      in-target: true    # defaults to false
```

in any case, I think we need to think about what it means in terms of backward compatibility

File: subiquity/server/controllers/tests/test_cmdlist.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+class TestEarlyController(TestCmdListController):
+    def setUp(self):
+        super().setUp()
+        self.controller = EarlyController(make_app())
+        self.controller.cmds = [Command(args="some-command", check=False)]
+
+    async def test_no_library_path_on_run(self):
+        with (
+            self.mocked_arun as faked_process_call,
```
Comment (ogayot) [inline subiquity/server/controllers/tests/test_cmdlist.py:44]: python3.8 is being annoying (hopefully we will get rid of it soon !) but you can work around the issue by dropping the parentheses :+1:

File: subiquitycore/tests/mocks.py [hunk 3]
```diff
@@ -43,4 +43,9 @@ def make_app(model=None):
     app.opts = mock.Mock()
     app.opts.dry_run = True
     app.scale_factor = 1000
+    app.echo_syslog_id = None
+    app.log_syslog_id = None
+    app.report_start_event = mock.Mock()
+    app.report_finish_event = mock.Mock()
```
Comment (ogayot) [inline subiquitycore/tests/mocks.py:49]: minor:

I think it's a useful change, thanks! Would you remove those now obsolete instructions (ideally in a separate commit along)?
```
subiquity/server/controllers/tests/test_filesystem.py:        self.app.report_start_event = mock.Mock()
subiquity/server/controllers/tests/test_filesystem.py:        self.app.report_start_event = mock.Mock()
subiquity/server/controllers/tests/test_install.py:        self.controller.app.report_start_event = Mock()
subiquity/server/controllers/tests/test_install.py:        self.controller.app.report_start_event = Mock()
subiquity/server/controllers/tests/test_refresh.py:        self.app.report_start_event = mock.Mock()
subiquity/server/controllers/tests/test_snaplist.py:        self.app.report_start_event = Mock()
```

**Review by ogayot — CHANGES_REQUESTED** [review]
LGTM from the code standpoint, thanks!

I am concerned about breaking the ability to run  `curtin in-target [...]` in late-commands though. I think we need to think about what it means for autoinstall users who are going to take a new subiquity snap.

File: subiquity/server/controllers/cmdlist.py (hunk 1 above)
Comment (ogayot) [reply subiquity/server/controllers/cmdlist.py:130]: Another option (which is not my favorite!) would be to check if the command starts with "curtin":

```python
if shlex.split(cmd.desc())[0] == "curtin":
    clean_env = False
```

Comment (dbungert) [reply subiquity/server/controllers/cmdlist.py:130]: I do think `shlex.split()` is where this needs to go.  Maybe we also pair that with `shutil.which()` and check if that resolves to something in the snap.

Comment (Chris-Peterson444) [reply subiquity/server/controllers/cmdlist.py:130]: Thanks for pointing this out. For the general case, I like the idea of inspecting the path from `shutil.which()` and determining if we should drop the snap environment variables or not.

Eventually adding support in the autoinstall config to specify the command environment seems like a good call, although perhaps we could encourage the use of `env` until we decide if want to and how that would look.

File: subiquitycore/tests/mocks.py (hunk 3 above)
Comment (Chris-Peterson444) [reply subiquitycore/tests/mocks.py:49]: Done! I went ahead and double checked any other instances of these instructions and it looks these were it.

File: subiquitycore/utils.py [hunk 4]
```diff
@@ -35,21 +35,24 @@ def _clean_env(env, *, locale=True):
...
         if key.endswith("_ORIG"):
-            key_to_restore = key[: -len("_ORIG")]
+            key_to_restore: str = key[: -len("_ORIG")]
             if val:
                 ret[key_to_restore] = val
             else:
                 del ret[key_to_restore]
             del ret[key]
-    ret.pop("LD_LIBRARY_PATH", None)
+    _ = ret.pop("LD_LIBRARY_PATH", None)
```
Comment (dbungert) [inline subiquitycore/utils.py:55]: The convention we're using in Subiquity is to not assign to underscore, as it's used for translation strings.  Do we need to assign to something?

File: subiquity/server/controllers/cmdlist.py [hunk 5]
```diff
@@ -78,6 +79,16 @@ async def run(self, context):
...
         for i, cmd in enumerate(tuple(self.builtin_cmds) + tuple(self.cmds)):
             desc = cmd.desc()
+
+            # If the path to the command isn't found on the snap we should
+            # drop the snap specific environment variables. If the command
+            # is `env`, we should let the user manage the environment
+            command = shlex.split(desc)[0]
+            path = shutil.which(command)
+            if path is not None:
+                if not path.startswith("/snap") and command != "env":
```
Comment (dbungert) [inline subiquity/server/controllers/cmdlist.py:89]: If we're special casing `env`, someone might have specified the full path, which means all of `("env", "/bin/env", "/usr/bin/env")` are possible.

File: subiquity/server/controllers/tests/test_cmdlist.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    @mock.patch("shutil.which", return_value="/usr/bin/path/to/bin")
+    async def test_no_snap_env_on_call(self, mocked_shutil):
+        with self.mocked_arun as faked_process_call, self.mocked_os_environ, self.mocked_orig_environ as mocked_orig_environ:
+            await self.controller.run()
+            _, kwargs = faked_process_call.call_args
+            call_env = kwargs["env"]
+
+            mocked_orig_environ.assert_called()
+            assert "LD_LIBRARY_PATH" not in call_env
```
Comment (dbungert) [inline subiquity/server/controllers/tests/test_cmdlist.py:53]: We're using the asserts that are available as test class members, generally.

```suggestion
            self.assertNotIn("LD_LIBRARY_PATH", call_env)
```

File: subiquitycore/utils.py (hunk 4 above)
Comment (Chris-Peterson444) [reply subiquitycore/utils.py:55]: Nope, this was just a personal preference when I was making edits.If that's convention I'll remove it.

File: subiquity/server/controllers/cmdlist.py (hunk 5 above)
Comment (Chris-Peterson444) [reply subiquity/server/controllers/cmdlist.py:89]: Hmmm, that's a good point. My thought was if someone passed their command via `env` they probably don't want us stripping everything for them (unless of course they're passing `env -i`). Should we special case on all three of these options, or just not special case at all?

Comment (dbungert) [reply subiquity/server/controllers/cmdlist.py:89]: /usr/bin/env does link to libc6, so the better answer is orig_env, I think.
for `env -i`  I think it doesn't matter what is done here.

> if someone passed their command via env they probably don't want us stripping everything for them

I would also prefer to not break someone's existing autoinstall, but if they are setting environment variables today I think the common scenarios are:
* they have a special case that is not covered by the above
* they already ran into this problem and are working around it, and giving a clean environment should still be ok

So I think the answer is to not special case env.  If you can think of a corner case not covered in this plan please point it out.

File: subiquity/server/controllers/tests/test_cmdlist.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        self.mocked_os_environ = mock.patch.dict("os.environ", snap_env_vars)
+
+    @mock.patch("shutil.which", return_value="/usr/bin/path/to/bin")
+    async def test_no_snap_env_on_call(self, mocked_shutil):
+        with self.mocked_arun as faked_process_call:
+            with self.mocked_orig_environ as mocked_orig_environ:
+                with self.mocked_os_environ:
+                    await self.controller.run()
+                    _, kwargs = faked_process_call.call_args
```
Comment (dbungert) [inline subiquity/server/controllers/tests/test_cmdlist.py:51]: more underscores

File: subiquity/server/controllers/tests/test_cmdlist.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                    mocked_orig_environ.assert_called()
+                    self.assertNotIn("LD_LIBRARY_PATH", call_env)
+
+    @mock.patch("shutil.which", return_value="/snap/path/to/bin")
+    async def test_with_snap_env_on_call(self, mocked_shutil):
+        with self.mocked_arun as faked_process_call:
+            with self.mocked_orig_environ as mocked_orig_environ:
+                with self.mocked_os_environ:
+                    await self.controller.run()
+                    _, kwargs = faked_process_call.call_args
```
Comment (dbungert) [inline subiquity/server/controllers/tests/test_cmdlist.py:63]: same

File: subiquity/server/controllers/cmdlist.py (hunk 5 above)
Comment (Chris-Peterson444) [reply subiquity/server/controllers/cmdlist.py:89]: I can't think of any other cases. That covers my concerns, so I'll remove the special casing. Thanks!

File: subiquity/server/controllers/tests/test_cmdlist.py (hunk 7 above)
Comment (Chris-Peterson444) [reply subiquity/server/controllers/tests/test_cmdlist.py]: Thanks for the catch, fixed.

File: subiquity/server/controllers/tests/test_cmdlist.py (hunk 8 above)
Comment (Chris-Peterson444) [reply subiquity/server/controllers/tests/test_cmdlist.py]: Also fixed.

File: subiquity/server/controllers/tests/test_cmdlist.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    CmdListController,
+    Command,
+    EarlyController,
+    LateController,
+)
+from subiquitycore.tests.mocks import make_app
+from subiquitycore.utils import orig_environ
+
+
+@mock.patch.dict("os.environ", {"LD_LIBRARY_PATH": "/var/lib/snapd/lib/gl"})
```
Comment (dbungert) [inline subiquity/server/controllers/tests/test_cmdlist.py:29]: I played with this a bit, is this just bugged on python3.8?

**Review by dbungert — APPROVED** [review]

**Comment by Chris-Peterson444** [comment]
I tried `curtin in-target -- touch /etc/test-file-here` and I can verify it shows up under `/target/etc/test-file-here`

File: subiquity/server/controllers/tests/test_cmdlist.py (hunk 9 above)
Comment (dbungert) [reply subiquity/server/controllers/tests/test_cmdlist.py:29]: I think this line is obsolete now?
